File: apps/kotlin/src/main/kotlin/chapter02/IrisPreprocessing.kt
package chapter02

import org.jetbrains.kotlinx.dataframe.AnyCol
import org.jetbrains.kotlinx.dataframe.AnyFrame
import org.jetbrains.kotlinx.dataframe.DataFrame
import org.jetbrains.kotlinx.dataframe.api.fillNulls
import org.jetbrains.kotlinx.dataframe.api.remove
import org.jetbrains.kotlinx.dataframe.api.with
import org.jetbrains.kotlinx.dataframe.io.readCSV
import java.io.File
import kotlin.math.ceil
import kotlin.random.Random

fun loadIris(csvFile: File): AnyFrame = DataFrame.readCSV(csvFile)

private fun AnyCol.countNulls(): Int = values().count { it == null }

fun countMissing(df: AnyFrame): Map<String, Int> = df.columns().associate { it.name() to it.countNulls() }

fun columnMeans(
    df: AnyFrame,
    columns: List<String>,
): Map<String, Double> =
    columns.associateWith { name ->
        df[name]
            .values()
            .filterIsInstance<Number>()
            .map { it.toDouble() }
            .average()
    }

fun fillMissing(
    df: AnyFrame,
    values: Map<String, Double>,
): AnyFrame = values.entries.fold(df) { filled, (name, value) -> filled.fillNulls(name).with { value } }

fun splitFeaturesAndTarget(
    df: AnyFrame,
    target: String,
): Pair<AnyFrame, List<String>> = df.remove(target) to df[target].values().map { it.toString() }

data class TrainTestSplit<T>(
    val xTrain: AnyFrame,
    val xTest: AnyFrame,
    val tTrain: List<T>,
    val tTest: List<T>,
)

fun <T> splitTrainTest(
    x: AnyFrame,
    t: List<T>,
    testSize: Double,
    seed: Int,
): TrainTestSplit<T> {
    val positions = (0 until x.rowsCount()).shuffled(Random(seed))
    val nTrain = x.rowsCount() - ceil(x.rowsCount() * testSize).toInt()
    val train = positions.take(nTrain)
    val test = positions.drop(nTrain)
    return TrainTestSplit(
        xTrain = x[train],
        xTest = x[test],
        tTrain = t.slice(train),
        tTest = t.slice(test),
    )
}

const val TARGET = "種類"

fun prepareIris(
    csvFile: File,
    testSize: Double,
    seed: Int,
): TrainTestSplit<String> {
    val (x, t) = splitFeaturesAndTarget(loadIris(csvFile), TARGET)
    val split = splitTrainTest(x, t, testSize, seed)
    val means = columnMeans(split.xTrain, x.columnNames())
    return split.copy(
        xTrain = fillMissing(split.xTrain, means),
        xTest = fillMissing(split.xTest, means),
    )
}

File: apps/python/lib/chapter02/iris_preprocessing.py
import math
from dataclasses import dataclass, replace
from pathlib import Path

import numpy as np
import pandas as pd

TARGET = "種類"


@dataclass(frozen=True)
class TrainTestSplit:
    x_train: pd.DataFrame
    x_test: pd.DataFrame
    t_train: pd.Series
    t_test: pd.Series


def load_iris(csv_file: Path) -> pd.DataFrame:
    return pd.read_csv(csv_file)


def count_missing(df: pd.DataFrame) -> pd.Series:
    return df.isna().sum()


def column_means(df: pd.DataFrame, columns: list[str]) -> dict[str, float]:
    return {column: float(df[column].mean()) for column in columns}


def fill_missing(df: pd.DataFrame, values: dict[str, float]) -> pd.DataFrame:
    return df.fillna(values)


def split_features_and_target(
    df: pd.DataFrame, target: str
) -> tuple[pd.DataFrame, pd.Series]:
    return df.drop(columns=[target]), df[target]


def split_train_test(
    x: pd.DataFrame, t: pd.Series, test_size: float, seed: int
) -> TrainTestSplit:
    positions = np.random.default_rng(seed).permutation(len(x))
    n_train = len(x) - math.ceil(len(x) * test_size)
    train, test = positions[:n_train], positions[n_train:]
    return TrainTestSplit(
        x_train=x.iloc[train],
        x_test=x.iloc[test],
        t_train=t.iloc[train],
        t_test=t.iloc[test],
    )


def prepare_iris(csv_file: Path, test_size: float, seed: int) -> TrainTestSplit:
    x, t = split_features_and_target(load_iris(csv_file), TARGET)
    split = split_train_test(x, t, test_size=test_size, seed=seed)
    means = column_means(split.x_train, list(x.columns))
    return replace(
        split,
        x_train=fill_missing(split.x_train, means),
        x_test=fill_missing(split.x_test, means),
    )
